src/net_processing.cpp

```diff
@@ -3330,6 +3330,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
         if (!vRecv.empty())
             vRecv >> fRelay;
+
+        // Don't waste any of our full outbound slots on peers that don't want tx relay (because they might be in -blocksonly mode)
+        if (!fRelay && !m_opts.ignore_incoming_txs && pfrom.IsFullOutboundConn())
+        {
+            LogPrint(BCLog::NET, "peer=%d (full-relay outbound) does not participate in transaction relay; disconnecting\n", pfrom.GetId());
```

Comment:
Rather than doing this immediaely upon connection (ie, after receiving their VERSION message), would it be better to do it only when we have all our outbound slots filled up (ie, as part of `EvictExtraOutboundPeers`) ?

Also, doing the check repeatedy would mean that we could keep blocks-only peers while we're doing IBD, only dropping them when we exit IBD and start caring about tx relay.

Reply:
> Rather than doing this immediaely upon connection (ie, after receiving their VERSION message), would it be better to do it only when we have all our outbound slots filled up (ie, as part of EvictExtraOutboundPeers) ?

I like that idea. If for some strange reason we wouldn't be able to find any other peers we wouldn't end up connecting the only ones we have. But I think we would want to start evicting if we are at the limit (so in `CheckForStaleTipAndEvictPeers`), not just if we have exceeded it (`EvictExtraOutboundPeers`).

> Also, doing the check repeatedy would mean that we could keep blocks-only peers while we're doing IBD, only dropping them when we exit IBD and start caring about tx relay.

Do we really want that? The main reason to run in `-blocksonly` mode is to save traffic. I would assume that most nodes wouldn't enable listening anyway, but those that do (maybe not even on purpose), maybe it would be nicer not to use them for IBD?

Reply:
>start evicting if we are at the limit (so in CheckForStaleTipAndEvictPeers), not just if we have exceeded it (EvictExtraOutboundPeers).

I agree.

>Do we really want that? The main reason to run in -blocksonly mode is to save traffic. I would assume that most nodes wouldn't enable listening anyway, but those that do (maybe not even on purpose), maybe it would be nicer not to use them for IBD?

Not sure about this. You seem to care specifically about those users which do `-blocksonly` and forget to disable inbounds. Perhaps this is better communicated through documentation and logging. I think we should leave them a chance to strengthen the block relaying network.

Reply:
> it seems odd / logically inconsistent to run a node in -blocksonly and enable inbounds, so I don't think it's worth overly optimizing for the use case

Maybe there is a use case for wanting to help the network by providing inbound capacity, but still minimize traffic at the same time? Although then it often doesn't make a lot of sense to run a non-pruned node - unless you also need the entire chain for selfish purposes (e.g. txindex).
In any case, I have included the suggestion to not evict while in ibd in the latest push!

Reply:
> Do we really want that? The main reason to run in `-blocksonly` mode is to save traffic.

In that case you'd be better off doing `-maxuploadtarget` either instead of or in addition to `-blocksonly`?

> it seems odd / logically inconsistent to run a node in `-blocksonly` and enable inbounds,

I think it would make sense to expect that if 100% (or 99.9%) of nodes were `-blocksonly` that they would still manage to propagate blocks sensibly, bring up new nodes efficiently, etc. In theory with this PR, we would also be able to propagate txs successfully in the 99.9% `-blocksonly` case.